File: cpp/graphs/dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> dijkstra(int V, vector<vector<int>> adj[], int S) {
  set<pair<int, int>> st;
  vector<int> dist(V, 1e9);

  st.insert({0, S});
  dist[S] = 0;

  while (!st.empty()) {
    auto it = *(st.begin());
    int node = it.second;
    int dis = it.first;
    st.erase(it);

    for (auto it : adj[node]) {
      int adjNode = it[0]; // Adjacent node
      int edgW = it[1];    // Weight of the edge
      if (dis + edgW < dist[adjNode]) {
        if (dist[adjNode] != 1e9)
          st.erase({dist[adjNode], adjNode});

        dist[adjNode] = dis + edgW;
        st.insert({dist[adjNode], adjNode});
      }
    }
  }
  return dist;
}
int main() {
  int V, E, S;
  cout << "Enter number of Vertices (V), Edges (E), and Source Node (S):"
       << endl;
  if (!(cin >> V >> E >> S))
    return 0;

  vector<vector<int>> adj[V];

  cout << "Enter " << E << " edges in format: u v weight" << endl;
  for (int i = 0; i < E; i++) {
    int u, v, w;
    cin >> u >> v >> w;
    adj[u].push_back({v, w});
  }

  vector<int> res = dijkstra(V, adj, S);

  cout << "\n--- Shortest Path Results ---" << endl;
  for (int i = 0; i < V; i++) {
    cout << "Path " << S << " -> " << i << " : ";
    if (res[i] == 1e9)
      cout << "Unreachable" << endl;
    else
      cout << res[i] << " units" << endl;
  }

  return 0;
}

File: python/graphs/dijkstra.py
import heapq


class Solution:

    # Function to find the shortest distance of all the vertices from
    # the source vertex S
    def dijkstra(self, V, adj, S):
        # Create a priority queue to store the nodes as a pair {dist, node}
        # The queue will always return the node with
        # the smallest distance first
        pq = []

        # Initialising dist list with a large number to indicate the nodes are
        # unvisited initially
        dist = [float('inf')] * V
        dist[S] = 0

        # Push the source node with distance 0 into the priority queue
        heapq.heappush(pq, (0, S))

        while pq:
            # Pop the node with the smallest distance
            dis, node = heapq.heappop(pq)

            # Traverse all the adjacent nodes
            for adjNode, weight in adj[node]:
                # If the new calculated distance is smaller, update it
                if dis + weight < dist[adjNode]:
                    dist[adjNode] = dis + weight
                    heapq.heappush(pq, (dist[adjNode], adjNode))

        return dist


# Driver code
if __name__ == "__main__":
    # Number of vertices and edges
    V = 3
    E = 3
    S = 2

    # Create adjacency list to represent the graph
    adj = [[] for _ in range(V)]

    # Add edges to the graph
    adj[0].append((1, 1))
    adj[0].append((2, 6))
    adj[1].append((2, 3))
    adj[1].append((0, 1))
    adj[2].append((1, 3))
    adj[2].append((0, 6))

    # Create an instance of the Solution class
    obj = Solution()
    # Call dijkstra function and store the result
    res = obj.dijkstra(V, adj, S)

    # Output the shortest distance from source to all nodes
    print(" ".join(map(str, res)))
